File: 0125-valid-palindrome/0125-valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        int l = 0, r = s.length() - 1;
        while (l < r){
            while (l < s.length() && isAlphaNum(s, l)) l++;
            while (r >= 0 && isAlphaNum(s, r)) r--;
            if (l < r){
                if (Character.toLowerCase(s.charAt(l)) != Character.toLowerCase(s.charAt(r))){
                    return false;
                }
                l++;
                r--;
            }
        }
        return true;
    }

    boolean isAlphaNum(String s, int idx){
        char c = s.charAt(idx);
        if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) 
            return false;
        return true;
    }
}

File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        def alphaNum(c):
            return ("A" <= c <= "Z") or ("a" <= c <= "z") or ("0" <= c <= "9")

        def lower(c):
            if "A" <= c <= "Z":
                return chr(ord(c) + 32)
            return c

        l, r = 0, len(s) - 1
        while l < r:
            while l < r and not alphaNum(s[l]):
                l += 1
            while l < r and not alphaNum(s[r]):
                r -= 1
            if lower(s[l]) != lower(s[r]):
                return False
            l += 1
            r -= 1
            
        return True
